kotori/daq/storage/cratedb.py

```diff
@@ -0,0 +1,272 @@
+# -*- coding: utf-8 -*-
+# (c) 2023 Andreas Motl <andreas@getkotori.org>
+import calendar
+import functools
+import json
+from collections import OrderedDict
+from decimal import Decimal
+from copy import deepcopy
+from datetime import datetime, date, timezone
+
+import crate.client.http
+import pytz
+import requests
+from crate import client
+from crate.client.converter import DefaultTypeConverter
+from crate.client.exceptions import ProgrammingError
+from funcy import project
+from munch import Munch
+from twisted.logger import Logger
+
+from kotori.daq.storage.util import format_chunk
+
+log = Logger()
+
+
+class CrateDBAdapter:
+    """
+    Kotori database backend adapter for CrateDB.
+
+    CrateDB is a distributed and scalable SQL database for storing and analyzing massive
+    amounts of data in near real-time, even with complex queries. It is PostgreSQL-compatible,
+    and based on Lucene.
+
+    https://github.com/crate/crate
+    """
+
+    def __init__(self, settings=None, database=None):
+        """
+        Carry over connectivity parameters.
+
+        TODO: Verify with CrateDB Cloud.
+        """
+
+        settings = deepcopy(settings) or {}
+        settings.setdefault("host", "localhost")
+        settings.setdefault("port", "4200")
+        settings.setdefault("username", "crate")
+        settings.setdefault("password", "")
+        settings.setdefault("database", database)
+
+        # TODO: Bring back pool size configuration.
+        # settings.setdefault('pool_size', 10)
+
+        settings["port"] = int(settings["port"])
+
+        # FIXME: This is bad style. Well, but it is currently
+        #        inherited from ~10 year old code, so c'est la vie.
+        self.__dict__.update(**settings)
+
+        # Bookkeeping for all databases having been touched already
+        self.databases_written_once = set()
+
+        self.host_uri = "{host}:{port}".format(**self.__dict__)
+
+        # TODO: Bring back pool size configuration.
+        # log.info('Storage target is {uri}, pool size is {pool_size}', uri=self.host_uri, pool_size=self.pool_size)
+        log.info("Storage target is {uri}", uri=self.host_uri)
+        self.db_client = client.connect(
+            self.host_uri, username=self.username, password=self.password, pool_size=20,
+        )
```

Comment:
Verify and demonstrate connecting also to CrateDB Cloud, maybe on behalf of a subsequent iteration.